catalog/imaging/imaging-transformer-chipping/src/main/java/org/codice/alliance/imaging/chip/transformer/AbstractJpegService.java

```diff
@@ -0,0 +1,156 @@
+ */
+package org.codice.alliance.imaging.chip.transformer;
+
+import java.awt.Graphics;
+import java.awt.image.BufferedImage;
+import java.awt.image.ColorModel;
+import java.awt.image.WritableRaster;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+import javax.imageio.stream.ImageInputStreamImpl;
+
+import org.codice.imaging.nitf.core.image.ImageCompression;
+import org.codice.imaging.nitf.core.image.ImageMode;
+import org.codice.imaging.nitf.core.image.ImageSegment;
+
+abstract class AbstractJpegService implements JpegService {
+
+    private BufferedImage copyImage(BufferedImage source) {
+
+        ColorModel colorModel = source.getColorModel();
+
+        int width = source.getWidth();
+        int height = source.getHeight();
+
+        WritableRaster raster = source.getRaster()
+                .createCompatibleWritableRaster(width, height);
+
+        BufferedImage bufferedImage = new BufferedImage(colorModel,
+                raster,
+                colorModel.isAlphaPremultiplied(),
+                null);
+
+        Graphics g = bufferedImage.getGraphics();
+        g.drawImage(source, 0, 0, null);
+        g.dispose();
+        return bufferedImage;
+    }
+
+    private ImageInputStreamImpl createImageInputStream(final byte[] jpegData) {
+        return new ImageInputStreamImpl() {
+
+            private InputStream inputStream = new ByteArrayInputStream(jpegData);
+
+            @Override
+            public int read() throws IOException {
+                return inputStream.read();
+            }
+
+            @Override
+            public int read(byte[] b, int off, int len) throws IOException {
+                return inputStream.read(b, off, len);
+            }
+        };
+    }
+
+    private void setImageDataLengthField(ImageSegment chipImageSegment, byte[] jpegData) {
+        chipImageSegment.setDataLength(jpegData.length);
+    }
+
+    private void setImageDataField(ImageSegment chipImageSegment, final byte[] jpegData) {
+        chipImageSegment.setData(createImageInputStream(jpegData));
+    }
+
+    abstract ImageCompression getImageCompressionType();
+
+    abstract String getImageCompressionRateString(double actualBpppb);
+
+    abstract ImageMode getImageMode();
+
+    @Override
+    public final void createJpeg(BufferedImage bufferedImage, ImageSegment chipImageSegment,
+            boolean isBlocking, int blockWidth, int blockHeight) throws IOException {
+
+        if (chipImageSegment.getNumBands() < 1) {
+            throw new IllegalStateException("the number of bands must be set");
+        }
+
+        BufferedImage copy = copyImage(bufferedImage);
+
+        byte[] jpegData;
+        if (isBlocking) {
+            jpegData = encodeBlocks(copy, blockWidth, blockHeight);
+        } else {
+            jpegData = encodeWholeImage(copy);
+        }
+
+        double bpppb = calculateActualBpppb(bufferedImage, chipImageSegment, jpegData);
```

Comment:
What is bpppb?  It may be helpful to have this unabbreviated for other developers